### ISSUE TITLE
Contrast and Brightness as Visual-attribute

### ISSUE THREAD

**Issue by neuromechanist** [hed-standard/hed-schemas#135]
While annotating the Healthy Brain Network project's Surround Suppression task, I found that `Opacity` and `Lunminence`,  `Hue` and `Saturation`, but I can't find `Contrast` and `Brightness`. While these values are potentially dependent, having the full range of definitions helps describe the events.

**Comment by neuromechanist** [comment]
Upon further investigation, it seems that `Luminance` is the correct word to use instead of brightness:
Here is the definition of Luminance and Brightness, as described in a [blog post](https://www.xrite.com/blog/luminance-and-brightness-what-is-the-difference) by _x.rite_, the color calibration tool from _Pantone_

**Luminance**: Luminance is the luminous intensity projected on a given area and direction. Luminance typically describes the intensity of emitted light.

**Brightness**: We perceive brightness when lumens fall on the rods and cones of our retina. When we speak of brightness, we use subjective words like “dim” or “brilliant” because brightness cannot be measured like luminance but can be scaled in percentages.

Here, I am not sure if we want to differentiate the difference between the two. My current take is to keep the luminance and not to include **Brightness** in the schema.

**Comment by neuromechanist** [comment]
Here is the definition of `Contrast`:

Contrast is the difference between the luminance of two objects. Either of the following equations can formalize the definition  of  contrast:

1. Micbelson contrast = (L<sub>max</sub> - L<sub>min</sub>) / (L<sub>max</sub> + L<sub>min</sub>)
where  _L<sub>max</sub>_ and  _L<sub>min</sub>_ are the maximum and minimum luminance of an image.

3. Weber contrast = (L - L<sub>back</sub>) / L<sub>back</sub>
where _L_  is  the  luminance  of  the  object  and _L<sub>back</sub>_  is  the  luminance  of  the  background

Source: 
1. Legge, G. E., D. H. Parish, A. Luebker, and L. H. Wurm. 1990. “Psychophysics of Reading. XI. Comparing Color Contrast and Luminance Contrast.” Journal of the Optical Society of America. A, Optics and Image Science 7 (10): 2002–10.
2. Rosa, Claudio, and Carlo Aleci. 2022. “Psychophysics in the Ophthalmological Practice—II. Contrast Sensitivity.” Annals of Eye Science 7 (1): 35–35.

**Comment by VisLab** [comment]
After HWG discussion --- it was agreed we should put in `Contrast` but not `Brightness` in the next release.  What should the parent node be?  

Also, formula 1 is a relative difference ... should that be in the description?  The definition also indicates two objects.  Should this be a Relation such as "Contrast-between".

Do the two references both give both of these definitions?

**Comment by dorahermes** [comment]
Contrast should be a Visual-attribute (not a Relation). 

Some examples: 
1) An image of e.g. a grating can have a certain contrast.
2) One image can have multiple contrasts, as contrast can be calculated in different ways (like those noted above, or in other ways) for a certain spatial resolution at different locations.

**Comment by neuromechanist** [comment]
Thanks a lot, @VisLab and @dorahermes, for your support.

> What should the parent node be?

I agree with @dorahermes that it should be under `Visual-attribute`.

> Formula 1 is a relative difference

Both formulas talk about differences in luminance. Contrast needs two objects to be defined. However, I think that _Contrast-between_ would cause unnecessary repetition. For example, for a _foreground_ gray circle and a white _background_, the short HED tag will look like:

(Visual-presentation, ((Foreground-view, (Circle, Gray)), (Background-view, White),  Contrast/0.5))

> Do the two references both give both of these definitions?
No, the Weber contrast is listed only in the second source. However, the first source is well-cited and clearly describes the effect of contrast on the subject performance. The second source covers how the contrast can be manipulated for psychophysical experiments.

**Comment by neuromechanist** [comment]
Talking to @smakeig, he emphasized that there are many contrasts, and what we are referring to here is _luminance contrast_.  The first source also emphasizes this point by comparing the attributes of color contrast and luminance contrast.

So, I wonder if it would be better to have `Luminance-contrast` under `Visual-attributes` to clarify the tag.

**Comment by dorahermes** [comment]
I agree that `Luminance-contrast` is clear and contrast-between does not work. Note that contrast can also be calculated using sets of filters in visual encoding models.

**Comment by smakeig** [comment]
I wonder, though, whether luminance-contrast, as we are seeking to use in
the Present movie annotation project, is not better thought of as a measure
applied to events rather than a response attribute per se...   Here, we are
using it as defined by a particular formula (not specified in the
HED schema, of course, as it has near-arbitrary parameters).  Further,
these parameters and the formula they belong to are not accepted
standards.  Compare, for example, the use of dB to specify the volume of a
sound - here, several dB scales are recognized (Db A, etc.).  Of course,
the same '?' might be raised about many other event attributes.  For
example, "color, 'very-dark-red'" does have an agreed formal definition, as
does 'latency, 21.344 sec'.  It is doubtless of use to enter e.g. computed
luminance contrast values in an events.tsv file column -- and HED tools
should be able to incorporate this into the HED string for (here) each HED
"movie-shot, Onset" event marker - to allow HED search to pick out events
with e.g." luminance-contrast > 3".

Perhaps HED should have a list of recommended measure column name labels
that are not 'at the level of' standard HED schema terms...

Else -  perhaps we should introduce a MovingImage library schema that
defines not only 'luminance-contrast' but also parameters entering into its
definition:
e.g.
   "luminance-ratio, 2.1777;
            defined by  npreceding-frames, 3; nfollowing-frames, 3;
luminance--measure, RMS; luminance-smoothing, mean"

The above would seem like 'overkill' ... as meaningful variations in this
definition are not likely to be represented in a data archive, certainly
not on the scale that would make their differences across datasets
meaningfully testable...

?

**Comment by smakeig** [comment]
I meant that comment to begin, "I wonder, though, whether
luminance-contrast, as we are seeking to use in the Present movie
annotation project, is not better thought of as a measure applied to events
rather than AN EVENT attribute per se..."

**Comment by neuromechanist** [comment]
@smakeig, we are not using)\, at least currently, `Luminance-contrast` for the Present movie. A good example of the use of this tag is the contrast between foreground and background in a couple of the Healthy Brain Network Project. Here is an example of the exemplary analysis of the Surround Suppression Paradigm from Langer et.al., Sci Data 2017:
<img width="693" alt="image" src="https://github.com/hed-standard/hed-schemas/assets/44906843/eb95dadc-e4c7-4cd7-90a9-7417b78db4a9">

The Luminance Ratio for the Present Movie (as briefly described under [BIDS issue 153](https://github.com/bids-standard/bids-specification/issues/153#issuecomment-1863769529) does not seem to be able to benefit from `Luminance-contrast,` because Log Luminance Ratio (LLR) is relating two separate frames (which is different from the `Luminance-contrast`, which requires the elements to be in the same scene).

This is not to discount the importance of annotating temporal context; I just don't think that `Luminance-contrast` can be extended for that 😊.

**Comment by neuromechanist** [comment]
To summarize, here is what we have discussed so far

Suggestion to include Luminance-contrast in the base schema:

**Tag**: `Luminance-contrast`

**Parent**: `Property/Sensory-property/Sensory-attribute/Visual-attribute` 

**Definition**: Luminance contrast is the difference between the luminance of two objects. The suggested quantification method is the relative difference measure of luminance: Contrast-luminance = (Lmax - Lmin) / (Lmax + Lmin), where Lmax and Lmin are the maximum and minimum luminance of [specific portions of] an image.

**Source**: Legge, G. E., D. H. Parish, A. Luebker, and L. H. Wurm. 1990. “Psychophysics of Reading. XI. Comparing Color Contrast and Luminance Contrast.” Journal of the Optical Society of America. A, Optics and Image Science 7 (10): 2002–10.

**Accept-value**: Yes, **Type** = (Non-negative, Fraction), **Range** = 0 to 1.

**Example**: Represent a gray circle over a white background with a contrast of 0.5.
```
    (Visual-presentation,
        (
            (Foreground-view, (Circle, Gray)), (Background-view, White), (Luminance-Contrast, Fraction/0.5)
        )
    )
```

**Comment by dorahermes** [comment]
This definition is incorrect in many cases and only applies to a subset of luminance contrast measurements (you are suggesting Michelson) and it can be calculated in multiple ways as mentioned above. I would therefore not give a 'suggested method for quantification', as there are many methods (Michelson, Weber, RMS, and other filtering approaches). Also note that the RMS contrast is not a fraction. I think the definition should be more general and allow future sub-tree extension with such specifics, e.g: 

Definition: Luminance contrast is the difference in luminance in [specific portions of] a scene or image. 

Accept-value: Yes, Type = (Non-negative), Range = 0 to 1.

**Comment by neuromechanist** [comment]
Thanks, @dorahermes,
Comparing the two definitions,
1. Luminance contrast is the difference between the luminance of two objects.
2. Luminance contrast is the difference in luminance in [specific portions of] a scene or image.

I don't think either of the definitions is incorrect. However, definition 1 is more general since it does not constrain `Luminance-contrast` to a scene or image.

There are indeed several ways for quantification of contrast (see Wikipedia page on [contrast](https://en.wikipedia.org/wiki/Contrast_(vision))). We should encourage the inclusion of the quantification method in the definition somehow. Also, in my opinion, providing a recommended method to quantify `Luminance-contrast` would encourage users to either use that definition or explicitly mention their quantification method. @VisLab, I wonder how we could recommend including the quantification method in the tag?

With a quick literature search, I observed that the Michelson definition seemed to be used often, so I added that as the suggested quantification method.

I agree that the type should not be `Fraction`, to include the RMS method for contrast quantification.

**Comment by dorahermes** [comment]
I have calculated contrast before in encoding models using sets of Gabor filters, and while it would be useful to have some suggestions, I do not think we should recommend any one in particular. Specific recommendations/luminance contrast types should be reserved for a level deeper.

**Comment by VisLab** [comment]
Type of value would be done with `SuggestedTag`.  Added as part of PR #140.

**Comment by VisLab** [comment]
Addressed in PR#149
